test/runtests.jl

```diff
@@ -1155,6 +1204,56 @@ Base.similar(bc::Broadcast.Broadcasted{Broadcast.ArrayStyle{MyArray}}, ::Type{El
 
     @test identity.(StructArray(x=StructArray(a=1:3)))::StructArray == [(x=(a=1,),), (x=(a=2,),), (x=(a=3,),)]
     @test (x -> x.x.a).(StructArray(x=StructArray(a=1:3))) == [1, 2, 3]
+    @test identity.(StructArray(x=StructArray(x=StructArray(a=1:3))))::StructArray == [(x=(x=(a=1,),),), (x=(x=(a=2,),),), (x=(x=(a=3,),),)]
+    @test (x -> x.x.x.a).(StructArray(x=StructArray(x=StructArray(a=1:3)))) == [1, 2, 3]
+
+    @testset "ambiguity check" begin
+        test_set = Any[StructArray([1;2+im]),
+                    1:2, 
+                    (1,2),
+                    StructArray(@SArray [1;1+2im]),
+                    (@SArray [1 2]),
+                    1]
+        tested_style = StructArrayStyle[]
+        dotaddsub((a, b, c),) = @. a + b - c
+        for is in Iterators.product(randperm(6), randperm(6), randperm(6))
```

Comment:
I'm not sure I understand this change (same holds for the use of `randperm(4)` above). `Iterators.product` will go over all combinations. Why not just

```julia
for as in Iterators.product(testset, testset, testset)
```

? Or even as you had it before,

```julia
for a in test_set, b in test_set, c in test_set
```

Reply:
With this we randomly pick some `a`, `b`, `c` to check the result of `dotaddsub`.
This should be more though than `for as in Iterators.product(testset, testset, testset)` as we now only test each dest style only once.

Reply:
I agree with the idea that one should only check the destination style once. What confuses me is why is `randperm(6)` needed over simply `1:6`? Is the idea that you prefer to randomly select triples `a, b, c` that give a specific style type (rather than always take the first triple that occurs with that style type)?

That might make sense, but at least it'd be best to add some `Random.seed!(1234)` at the beginning to make it reproducible, otherwise I think it's going to be hard to debug failing tests in the future if what exactly is tested varies.

Reply:
Make sense.
A fixed seed seems no better than `for as in Iterators.product(testset, testset, testset)`.
After a second thought, a thorough style test should be enough (and much cheaper).
The latest version fully drop the usage of `randperm`s